### PR TITLE
feat: add partitions

### PR DESCRIPTION
**Pull request by fraidev**
Needs:
- [x] https://github.com/infinyon/fluvio/pull/4076

### CODE REVIEW

File: crates/fluvio-cli/src/client/topic/add_partition.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        let request = AddPartition {
+            count: self.count as u32,
+        };
+
+        let action = UpdateTopicAction::AddPartition(request);
+        let topic_before = admin
+            .list::<TopicSpec, _>(vec![self.topic.clone()])
+            .await?
+            .first()
+            .expect("topic not found")
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs]: Removed! Using anyhow instead, WDYT?

File: crates/fluvio-cli/src/client/topic/add_partition.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            .clone();
+        let replica_map_before = topic_before.status.replica_map;
+
+        let mut partition_stream = admin.watch::<TopicSpec>().await?;
+        let timeout = sleep(std::time::Duration::from_millis(
+            CHECK_NEW_PARTITIONS_TIMEOUT_MS,
+        ));
+
+        let send_new_partitions_and_watch_them = async {
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs]: I created a new method for the AddPartitionOpt with this logic, WDYT?

File: crates/fluvio-cli/src/client/topic/add_partition.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            _ = timeout => {
+                println!("response timeout exceeded");
+            },
+        }
+
+        Ok(())
+    }
+}
+
+fn display(diff: Vec<(u32, Vec<i32>)>, topic: &str) {
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs]: I changed to return String and added tests

File: crates/fluvio-sc/src/services/public_api/topic/update/add_partition.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
+//!
+//! # Add Partition Request
+//!
+use tracing::instrument;
+use std::io::Error;
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-sc/src/services/public_api/topic/update/add_partition.rs:4]: Done

File: crates/fluvio-sc/src/services/public_api/topic/update/mod.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
+mod add_partition;
+
+use std::io::{Error, ErrorKind};
+
+use anyhow::Result;
+
+use fluvio_controlplane_metadata::extended::SpecExt;
+use fluvio_protocol::link::ErrorCode;
+use tracing::{info, instrument, trace};
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-sc/src/services/public_api/topic/update/mod.rs]: Done

File: crates/fluvio-sc/src/services/public_api/topic/update/add_partition.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        let mut status = topic.status().clone();
+        status.resolution = TopicResolution::Pending;
+
+        auth_ctx
+            .global_ctx
+            .topics()
+            .create_spec(topic.key.clone(), spec)
+            .await?;
+
+        auth_ctx
```
[the remark this answers is not in the document]
Comment (fraidev) [reply crates/fluvio-sc/src/services/public_api/topic/update/add_partition.rs]: Oh, that makes a lot of sense! 
I moved this to the reducer of the topic controller.

File: crates/fluvio-cli/src/client/topic/add_partition.rs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    ) -> Result<ReplicaMap> {
+        let request = AddPartition {
+            count: self.count as u32,
+        };
+
+        let action = UpdateTopicAction::AddPartition(request);
+        let mut partition_stream = admin.watch::<TopicSpec>().await?;
+        let mut connected = false;
+        loop {
+            if let Some(Ok(event)) = partition_stream.next().await {
```
Comment (sehz) [inline crates/fluvio-cli/src/client/topic/add_partition.rs:86]: add timer so this doesn't result in infinite loop

Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs:86]: The timer outside does not cover it? 

https://github.com/infinyon/fluvio/pull/4065/files/48d0c616794f8f53f31c2ecf15fc41cddc6eb0be#diff-5660ec23133d43cf7faef763ee6f34d821c87a1ac45f4bb74f8a1e4154ae4f4cR65

Comment (sehz) [reply crates/fluvio-cli/src/client/topic/add_partition.rs:86]: just move timer inside then can get rid of extras select in the outside

File: crates/fluvio-cli/src/client/topic/add_partition.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                if !connected {
```
Comment (sehz) [inline crates/fluvio-cli/src/client/topic/add_partition.rs:91]: timer should be insert here

Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs]: Added!

File: crates/fluvio-cli/src/client/topic/add_partition.rs (hunk 7 above)
Comment (fraidev) [reply crates/fluvio-cli/src/client/topic/add_partition.rs]: Done, WDYT?

**Review by sehz — APPROVED** [review]
LGTM

**Pull request merged**